File: python/atlas_citations/scan.py
"""Read the handoff file that TypeScript writes.

``task:0029`` D1 puts the language boundary at ``data/citations/citations.json``:
TypeScript walks the AST and emits every citation instance with its location;
everything downstream of that is Python. This module is the Python side of that
contract and the only place that knows the file's shape.

The dataclasses here mirror ``cli/commands/citations/scan.ts``. Keeping them
explicit rather than passing raw dicts around is what makes a shape change fail
loudly at the boundary instead of quietly producing an empty bibliography three
commands later.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

#: Relative to the repo root.
SCAN_PATH = Path("data") / "citations" / "citations.json"

#: The only schema this reader understands. TypeScript bumps it on a shape change.
SCHEMA_VERSION = 1


class ScanError(RuntimeError):
    """The handoff file is missing, unreadable, or of an unknown schema version."""


@dataclass(frozen=True)
class Citation:
    """One citation instance, as extracted from the documents."""

    #: Canonical URL — the bibliography entry identity. ``None`` for ``unlinked``.
    key: str | None
    #: The URL exactly as it appeared, kept so a report can show the original.
    raw_url: str | None
    #: The link text, or the matched text for an unlinked citation.
    anchor_text: str
    #: ``citation`` | ``content-link`` | ``asset`` | ``unlinked``
    kind: str
    #: ``inline`` | ``footnote``
    origin: str
    footnote_number: str | None
    chapter_number: int
    section_number: int
    section_slug: str
    #: Parsed anchor text, or ``None`` when it does not split into author and year.
    #:
    #: Computed on the TypeScript side deliberately: ``author-year.ts`` holds the
    #: single definition of a citation anchor, shared with the audio renderer so
    #: the two cannot drift (``task:0025`` AC-6). Reimplementing that pattern here
    #: would fork it.
    author_year: dict[str, str] | None

    @property
    def location(self) -> str:
        """``ch1.2`` plus the footnote number when there is one."""
        where = f", footnote {self.footnote_number}" if self.origin == "footnote" else ""
        return f"ch{self.chapter_number}.{self.section_number}{where}"

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> Citation:
        return cls(
            key=raw.get("key"),
            raw_url=raw.get("rawUrl"),
            anchor_text=raw.get("anchorText", ""),
            kind=raw.get("kind", "content-link"),
            origin=raw.get("origin", "inline"),
            footnote_number=raw.get("footnoteNumber"),
            chapter_number=raw.get("chapterNumber", 0),
            section_number=raw.get("sectionNumber", 0),
            section_slug=raw.get("sectionSlug", ""),
            author_year=raw.get("authorYear"),
        )


@dataclass(frozen=True)
class Section:
    number: int
    title: str
    slug: str
    citations: list[Citation] = field(default_factory=list)


@dataclass(frozen=True)
class Chapter:
    number: int
    title: str
    slug: str
    sections: list[Section] = field(default_factory=list)


@dataclass(frozen=True)
class Scan:
    chapters: list[Chapter]

    def all_citations(self) -> list[Citation]:
        return [c for ch in self.chapters for s in ch.sections for c in s.citations]


def parse_scan(text: str) -> Scan:
    """Parse the handoff file. Raises :class:`ScanError` on anything unexpected."""
    try:
        raw = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ScanError(f"citations.json is not valid JSON: {exc}") from exc
    if not isinstance(raw, dict):
        raise ScanError("citations.json must contain an object")

    version = raw.get("schemaVersion")
    if version != SCHEMA_VERSION:
        raise ScanError(
            f"citations.json is schema version {version!r}, expected {SCHEMA_VERSION} — "
            "re-run `atlas citations scan`"
        )

    return Scan(
        chapters=[
            Chapter(
                number=ch.get("number", 0),
                title=ch.get("title", ""),
                slug=ch.get("slug", ""),
                sections=[
                    Section(
                        number=sec.get("number", 0),
                        title=sec.get("title", ""),
                        slug=sec.get("slug", ""),
                        citations=[Citation.from_json(c) for c in sec.get("citations", [])],
                    )
                    for sec in ch.get("sections", [])
                ],
            )
            for ch in raw.get("chapters", [])
        ]
    )


def read_scan(root: Path) -> Scan:
    """Read the handoff file from a checkout root."""
    path = root / SCAN_PATH
    if not path.exists():
        raise ScanError(f"no citation scan at {SCAN_PATH} — run `atlas citations scan` first")
    return parse_scan(path.read_text(encoding="utf-8"))

File: cli/commands/citations/scan.ts
/**
 * `atlas citations scan` - the TypeScript half of the citation pipeline, and
 * the only thing that crosses into Python.
 *
 * `task:0029` D1 puts the language boundary here. Extraction stays in
 * TypeScript because it walks the AST that 1,123 lines of loader code produce,
 * and that AST is not callable from outside this process. Everything after it -
 * resolvers, the CSL store, BibTeX, the reports - is ordinary scraping and
 * bibliography work with no tie to this language, and lives in `python/`.
 *
 * This command writes `data/citations/citations.json`: every citation instance
 * with its location, and nothing else. It is a handoff file, deliberately
 * inspectable, so a contributor can read exactly what passes between the two
 * halves instead of inferring it.
 *
 * One field here is not in `CitationInstance` and is the reason this file exists
 * rather than a straight JSON dump: `authorYear`. `author-year.ts` holds the
 * single definition of what a citation anchor looks like, shared with the audio
 * renderer so the two cannot drift (`task:0025` AC-6). The Python side needs the
 * parsed author and year for its report, and reimplementing the pattern there
 * would fork that definition - the precise failure AC-6 exists to prevent. So
 * the parse is performed once, here, and its result travels in the file.
 */
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import type { Chapter } from '../../../src/textbook-loader/index.js';
import { splitAuthorYear } from '../../../src/textbook-loader/citations/author-year.js';
import {
  extractSectionCitations,
  type CitationInstance,
} from '../../../src/textbook-loader/citations/extract.js';
import { loadChaptersFromCache } from './load.js';

/** Where the handoff file lives, relative to the repo root. */
export const SCAN_PATH = join('data', 'citations', 'citations.json');

/** A citation instance plus the author-year parse, which only TypeScript can make. */
export type ScannedCitation = CitationInstance & {
  /** Parsed anchor text, or null when it does not split into author and year. */
  authorYear: { author: string; year: string } | null;
};

export type ScannedSection = {
  number: number;
  title: string;
  slug: string;
  citations: ScannedCitation[];
};

export type ScannedChapter = {
  number: number;
  title: string;
  slug: string;
  sections: ScannedSection[];
};

export type ScanFile = {
  /** Bumped when the shape changes, so the Python side can refuse an old file. */
  schemaVersion: 1;
  chapters: ScannedChapter[];
};

/** Build the handoff structure. Pure: chapters in, plain data out. */
export function buildScan(chapters: Chapter[]): ScanFile {
  return {
    schemaVersion: 1,
    chapters: chapters.map((chapter) => ({
      number: chapter.number,
      title: chapter.title,
      slug: chapter.slug || `chapter-${chapter.number}`,
      sections: chapter.sections.map((section) => ({
        number: section.number,
        title: section.title,
        slug: section.slug ?? '',
        citations: extractSectionCitations(section).map((c) => ({
          ...c,
          authorYear: splitAuthorYear(c.anchorText),
        })),
      })),
    })),
  };
}

/** Run the command: load the cached chapters, write the handoff file. */
export async function citationsScan(root: string, outPath?: string): Promise<number> {
  const chapters = await loadChaptersFromCache();
  const scan = buildScan(chapters);

  const dest = outPath ?? join(root, SCAN_PATH);
  mkdirSync(dirname(dest), { recursive: true });
  writeFileSync(dest, JSON.stringify(scan, null, 2) + '\n', 'utf8');

  const total = scan.chapters.reduce(
    (n, c) => n + c.sections.reduce((m, s) => m + s.citations.length, 0),
    0,
  );
  console.log(`${total} citation instances across ${scan.chapters.length} chapters → ${dest}`);
  return 0;
}
